File: Javascript/Arrays/D2_MoveAllZeroesToEnd.js
function pushZeroesToEnd(arr) {
      let index = 0;

      for (let i = 0; i < arr.length; i++) {
            if (arr[i] !== 0) {
                  [arr[i], arr[index]] = [arr[index], arr[i]]; // Swap using destructuring
                  index++;
            }
      }
}

let arr1 = [1, 2, 0, 4, 3, 0, 5, 0];
let arr2 = [10, 20, 30];
let arr3 = [0, 0];

pushZeroesToEnd(arr1);
pushZeroesToEnd(arr2);
pushZeroesToEnd(arr3);

console.log(arr1);
console.log(arr2);
console.log(arr3);

File: Java/Arrays/D2_MoveAllZeroesToEnd.java
package Arrays;

import java.util.Arrays;


public class D2_MoveAllZeroesToEnd {
    public static void main(String[] args) {
        int[] arr1 = {1, 2, 0, 4, 3, 0, 5, 0};
        int[] arr2 = {10, 20, 30};
        int[] arr3 = {0, 0};

        pushZeroesToEnd(arr1);
        pushZeroesToEnd(arr2);
        pushZeroesToEnd(arr3);

        System.out.println(Arrays.toString(arr1));
        System.out.println(Arrays.toString(arr2));
        System.out.println(Arrays.toString(arr3));
    }
    public static void pushZeroesToEnd(int[] arr) {
        int left = 0, right = arr.length - 1;

        while (left < right) {
            while (left < right && arr[right] == 0) {
                right--;
            }

            if (arr[left] == 0) {
                swap(arr, left, right);
                right--;
            }
            left++;
        }
//        int nonZeroIndex = 0;
//
//        for (int i = 0; i < arr.length; i++) {
//            if (arr[i] != 0) {
//                swap(arr, i, nonZeroIndex);
//                nonZeroIndex++;
//            }
//        }
    }

    private static void swap(int[] arr, int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }
}
